### PR TITLE
Drop `rfc3986` requirement.

### LINKED ISSUE #1833: Some URLs with unescaped query strings cannot resolve relative redirects.
### Discussed in https://github.com/encode/httpx/discussions/1816

<div type='discussions-op-text'>

<sup>Originally posted by **kangzhang** August 25, 2021</sup>
Hey there, we are running into a crash on httpx and wondering if we could get some help. 

It seems rfc3986 is confused by certain URLs. To reproduce the bug:

```
from httpx import Client
Client().get("https://reason.com/search/?f[author][]=nick-gillespie&q=*&s=-pubdate")
```

It would throw a ResolutionError error about the URL.

Looking into the code, I'm wondering if we should patch this line and also set `query` to `None`: https://github.com/encode/httpx/blob/master/httpx/_models.py#L572

```
base_uri = self._uri_reference.copy_with(fragment=None, query=None)
```

The raw exception message:

```
  File "httpx/_client.py", line 1776, in head
    return await self.request(
  File "httpx/_client.py", line 1481, in request
    response = await self.send(
  File "httpx/_client.py", line 1568, in send
    response = await self._send_handling_auth(
  File "httpx/_client.py", line 1604, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "httpx/_client.py", line 1658, in _send_handling_redirects
    raise exc
  File "httpx/_client.py", line 1647, in _send_handling_redirects
    request = self._build_redirect_request(request, response)
  File "httpx/_client.py", line 456, in _build_redirect_request
    url = self._redirect_url(request, response)
  File "httpx/_client.py", line 508, in _redirect_url
    url = request.url.join(url)
  File "httpx/_models.py", line 573, in join
    return URL(relative_url._uri_reference.resolve_with(base_uri).unsplit())
  File "rfc3986/_mixin.py", line 266, in resolve_with
    raise exc.ResolutionError(base_uri)
```

Thank you!</div>

The shortest way to reproduce this issue is something like...

```python
>>> httpx.URL("https://example.com/?[]").join("/")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tomchristie/GitHub/encode/httpx/httpx/_models.py", line 574, in join
    return URL(relative_url._uri_reference.resolve_with(base_uri).unsplit())
  File "/Users/tomchristie/GitHub/encode/httpx/venv/lib/python3.6/site-packages/rfc3986/_mixin.py", line 266, in resolve_with
    raise exc.ResolutionError(base_uri)
rfc3986.exceptions.ResolutionError: https://example.com/?[] is not an absolute URI.
```

The REF3986 package is using a regex to determine if a URL is in absolute form or not, and the query part does not match the regex. I think @kangzhang's suggestion to remove the query portion when joining URLs would resolve this, and *looks* like it still would give correctly resolved redirects, but it'd be worth making the change, and seeing if anything in the test suite changes as a result of the change or not.

### LINKED ISSUE #2169: RFE: provide suport for `rfc3986` 2.0.0
Currently `rfc3986` 2.0.0 is not supported.
https://github.com/encode/httpx/blob/3af5146788f2945c806d4225cd588a2aa8073b90/setup.py#L58-L62
Do you have any plans to provide such support?

### PR DESCRIPTION
**Pull request by tomchristie**
This pull request drops the `rfc3986` package dependancy, in favour of a carefully worked through `urlparse` implementation, that provides everything we need in terms of URL validation and normalisation.

Closes #1833 (Fixed)
Closes #2169 (No longer required)
Closes #2175 (No longer required)

Marking this up as ready-for-review now.

Although this adds some additional code, it also completely removes a dependancy, and is lower overall-complexity. In terms of reading through and being able to understand the split between the `URL` model and the underlying URL parsing code, I think it ends up actually being much simpler.

I think the `rfc3986` package is fantastic, but we were having to do a few bits in some slightly kludgy ways to use it, and without the indirection I find it much clearer to work all the way through what's actually going on here.

Would be very happy to work through a review here step by step until we're confident that:

* We've ensured that the parsing here is (relatively) easy enough for someone else to work through.
* We've ensured that it's throughly enough source-documented.
* We've ensured that any concerns about confidence-in-correctness have been thoroughly addressed.

Potentially contentious, because because of the "rely on existing packages" argument, but I'd like to work through this with someone else. To my eyes it's actually a nice rationalisation. But then I've spent the time working on it, so?

### CODE REVIEW

File: tests/models/test_url.py
Comment (tomchristie) [inline tests/models/test_url.py:316]: Our test here is much simpler, since this URL no longer passes validation. 👍

File: httpx/_urls.py
Comment (tomchristie) [inline httpx/_urls.py:130]: Unlike with `rfc3986`, this value can no longer be `None`.

(It's not needed, since an empty string is sufficient.)

File: httpx/_urls.py
Comment (tomchristie) [inline httpx/_urls.py:146]: Unlike with `rfc3986`, this value can no longer be `None`.

(It's not needed, since an empty string is sufficient.)

File: httpx/_urls.py
Comment (tomchristie) [inline httpx/_urls.py:336]: Note that our parameter checking moves into `__init__(...)` instead.

File: httpx/_urls.py
Comment (tomchristie) [inline httpx/_urls.py:362]: We're just leaning on the stdlib's built-in implementation of `urljoin` now, but making sure to use our validation and normalisation first.

**Comment by tomchristie** [comment]
I might end up putting on my admin hat and pushing this one through.
Anyone interested is welcome to review it.

**Comment by adriangb** [comment]
I looked over the change. I agree that this implementation is very clean and easy to grok at first glance. Given httpx is an HTTP client, parsing URLs seems like a pretty fundamental behavior and I am not surprised that we benefit from doing it internally instead of relying on a 3rd party package. I'd be a lot more concerned if this was an app or library that only tangentially does HTTP stuff.

All this said, what I cannot evaluate is the correctness: it looks good at a glance, but I don't know the ins and outs of URL parsing / RFCs in the area, so for that I would have to defer to someone with more experience.

**Comment by xkortex** [comment]
Very nice work! I've been looking through the code over time and letting it soak in. Only thing that jumped out at me right away is there is no "authority" field. This field is mentioned fairly prominently in rfc3986, so IMHO it's worth having as a field/`@property`. Easy to add. 

I ran these through some examples in my [URL/URI torture test](https://github.com/xkortex/uri_torture_test) I've been working on, and collecting examples for around the net, and found a couple of discrepancies that I think are worth looking at. 

In particular, it seems to struggle with rfc4151 "tag" scheme URIs: 
```
<!> Failure case: rfc='rfc4151' item['uri']='tag:timothy@hpl.hp.com,2001:web/externalHome'
 key='path', value_want='web/externalHome' httpx_got='timothy@hpl.hp.com,2001:web/externalHome'

<!> Failure case: rfc='rfc4151' item['uri']='tag:sandro@w3.org,2004-05:Sandro'
 key='path', value_want='Sandro' httpx_got='sandro@w3.org,2004-05:Sandro'

<!> Failure case: rfc='rfc4151' item['uri']='tag:blogger.com,1999:blog-555'
 key='path', value_want='blog-555' httpx_got='blogger.com,1999:blog-555'
 ```

rfc4151 URIs are a little weird so I'm not surprised it is parsing them differently, but I figured I would point it out. 

There's also a batch of valid urls ([taken from here](https://mathiasbynens.be/demo/url-regex)) that don't parse into IDNA hostname correctly (throws Invalid IDNA hostname error), and some invalid URIs that ought to be throwing some kind of error, but nonetheless pass.

File: httpx/_urlparse.py
Comment (xkortex) [inline httpx/_urlparse.py:57]: I just want to point out, I love love LOVE using format strings to improve readability and maintainability of regex.

File: httpx/_urls.py
Comment (xkortex) [inline httpx/_urls.py:384]: Question: could this just be str(self._url_reference)? Just curious what the implications are of the different rendering methods are. I think it'd be more parsimonious to have a single URL->str but I could see this being a backwards-compat thing.

Comment (tomchristie) [reply httpx/_urls.py:384]: Right now we're doing this here because we need to mask the password for `__repr__`, but not for the `__str__`.

We could perhaps simplify a little by having a `to_string(mask_password: bool)` method on the `ParseResult` class.

Comment (xkortex) [reply httpx/_urls.py:384]: Ohhhh right, because of this, right? https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1

> We could perhaps simplify a little by having a to_string(mask_password: bool) method on the ParseResult class.

That also sounds reasonable, if it were my codebase, I'd prefer that. Not a big deal either way.

**Comment by xkortex** [comment]
So with those odd URIs, they do have parity with `rfc3986` library, which is probably what we want for this refactor, as we want it to be as transparent as possible. On the whole, it looks like this refactor does exactly that, which is excellent. Couch that in, "URIs/URLs are hard and there are millions of corner cases" but I think this does a great job at preserving the `rfc3986`-ness while adding much more flexibility. 

Seems the only discrepancy now is merely between this implementation and the WHATWG "living URL standard" (as an aside, I'm personally not even convinced that spec is the One True Spec, but that's a holy war for another day). Especially with some of the IDNA stuff, I think handling of unicode URLs could be a bit more graceful. E.g. the `rfc3986` lib makes a distinction between URI and IRI, and I don't think that's been pertinent in over a decade. I would expect `httpx-1.0` to be able to take a unicode URL and just know the correct thing to do with it. Sounds like a great opportunity for a follow up PR though!

File: httpx/_urls.py
Comment (tomchristie) [reply httpx/_urls.py:384]: > I think it would be better if it didn't use square braces to elide the password, since that screws up later parses, but that's beyond the purview of this PR

Hrm, yes, we could certainly reconsider that. As you say tho, we'd do so independently of this pull request.

**Comment by tomchristie** [comment]
> So with those odd URIs, they do have parity with rfc3986 library, which is probably what we want for this refactor

Exactly so, yes.

What do we think. Are we happy enough to pull this PR in as-is, or would we like to make things even tighter with a bunch of additional URL tests?

**Comment by tomchristie** [comment]
Okay I'm confident on this one.
Anyone up for a review?

**Comment by kloczek** [comment]
Looks like this PR is not possoible to apply on top of currert master
```console
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/python-httpx-Drop-rfc3986-requirement.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch -f
2 out of 2 hunks FAILED -- saving rejects to file httpx/_models.py.rej
1 out of 1 hunk FAILED -- saving rejects to file httpx/_types.py.rej
3 out of 3 hunks FAILED -- saving rejects to file httpx/_urls.py.rej
1 out of 1 hunk FAILED -- saving rejects to file tests/client/test_proxies.py.rej
1 out of 1 hunk FAILED -- saving rejects to file tests/models/test_url.py.rej
1 out of 1 hunk FAILED -- saving rejects to file setup.py.rej
```

**Comment by kloczek** [comment]
Any update about integrate thos PR and make new release?

**Comment by tomchristie** [comment]
@kloczek Yes, I'd be keen on that.
We're currently waiting on an approving review here.
I've sent you an invite to the maintainers team, which gives you review permissions.
I'd be happy to work on resolving any blockers we have on getting this resolved.

**Comment by kloczek** [comment]
> I've sent you an invite to the maintainers team, which gives you review permissions.

Maybe I'm looking in wrong place but I dont see anything like possibility to aprove something (and I did not receive any notyfication about that). 🤔

**Comment by kloczek** [comment]
Just back to `httpx` after 0.23.1 has been released and looks like this PR needs to be updated again
```console
+ /usr/bin/gzip -dc /home/tkloczko/rpmbuild/SOURCES/python-httpx-0.23.1.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd httpx-0.23.1
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ /usr/bin/cat /home/tkloczko/rpmbuild/SOURCES/python-httpx-Drop-rfc3986-requirement.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch -f
2 out of 2 hunks FAILED -- saving rejects to file httpx/_models.py.rej
1 out of 1 hunk FAILED -- saving rejects to file httpx/_types.py.rej
3 out of 3 hunks FAILED -- saving rejects to file httpx/_urls.py.rej
1 out of 1 hunk FAILED -- saving rejects to file tests/client/test_proxies.py.rej
1 out of 1 hunk FAILED -- saving rejects to file tests/models/test_url.py.rej
1 out of 1 hunk FAILED -- saving rejects to file setup.py.rej
```

**Comment by kloczek** [comment]
Tested again this PR against current maaster and there is no issues with rejections however pytest fails
<details>

```console
+ SETUPTOOLS_SCM_PRETEND_VERSION=0.23.1
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-httpx-0.23.1-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-httpx-0.23.1-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra -p no:randomly
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.16, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/httpx-0.23.1, configfile: setup.cfg
plugins: anyio-3.6.2, asyncio-0.20.2, trio-0.7.0
asyncio: mode=strict
collected 708 items

tests/test_api.py ...........                                                                                                                                        [  1%]
tests/test_asgi.py ....................                                                                                                                              [  4%]
tests/test_auth.py ....                                                                                                                                              [  4%]
tests/test_config.py ...............................                                                                                                                 [  9%]
tests/test_content.py .....................                                                                                                                          [ 12%]
tests/test_decoders.py ..............................                                                                                                                [ 16%]
tests/test_exceptions.py ....                                                                                                                                        [ 17%]
tests/test_exported_members.py .                                                                                                                                     [ 17%]
tests/test_main.py F
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 270, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 324, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 349, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 112, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 131, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 226, in call_and_report
INTERNALERROR>     hook.pytest_exception_interact(node=item, call=call, report=report)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pytest_trio/plugin.py", line 62, in pytest_exception_interact
INTERNALERROR>     if issubclass(call.excinfo.type, trio.MultiError):
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/trio/_deprecate.py", line 120, in __getattr__
INTERNALERROR>     warn_deprecated(thing, info.version, issue=info.issue, instead=instead)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/trio/_deprecate.py", line 51, in warn_deprecated
INTERNALERROR>     warnings.warn(TrioDeprecationWarning(msg), stacklevel=stacklevel)
INTERNALERROR> trio.TrioDeprecationWarning: trio.MultiError is deprecated since Trio 0.22.0; use BaseExceptionGroup (on Python 3.11 and later) or exceptiongroup.BaseExceptionGroup (earlier versions) instead (https://github.com/python-trio/trio/issues/2211)
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.8/logging/__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/usr/lib64/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib64/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib64/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 81, in serve
    await self.shutdown(sockets=sockets)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 253, in shutdown
    logger.info("Shutting down")
Message: 'Shutting down'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.8/logging/__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/usr/lib64/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib64/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib64/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 85, in serve
    logger.info(message, process_id, extra={"color_message": color_message})
Message: 'Finished server process [%d]'
Arguments: (2755761,)
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.8/logging/__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/usr/lib64/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib64/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib64/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 81, in serve
    await self.shutdown(sockets=sockets)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 253, in shutdown
    logger.info("Shutting down")
Message: 'Shutting down'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib64/python3.8/logging/__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/usr/lib64/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib64/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib64/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/uvicorn/server.py", line 85, in serve
    logger.info(message, process_id, extra={"color_message": color_message})
Message: 'Finished server process [%d]'
Arguments: (2755761,)

====================================================================== 1 failed, 122 passed in 3.52s =======================================================================
```
<details>

**Comment by tomchristie** [comment]
Up to date with master.

**Comment by kloczek** [comment]
> Up to date with master.

Are you sure?
```console
[tkloczko@devel-g2v httpx]$ git pull
Already up to date.
[tkloczko@devel-g2v httpx]$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
[tkloczko@devel-g2v httpx]$ wget https://github.com/encode/httpx//pull/2252.patch#/python-httpx-Drop-rfc3986-requirement.patch
--2022-12-09 10:57:02--  https://github.com/encode/httpx//pull/2252.patch
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://patch-diff.githubusercontent.com/raw/encode/httpx/pull/2252.patch [following]
--2022-12-09 10:57:02--  https://patch-diff.githubusercontent.com/raw/encode/httpx/pull/2252.patch
Resolving patch-diff.githubusercontent.com (patch-diff.githubusercontent.com)... 140.82.121.4
Connecting to patch-diff.githubusercontent.com (patch-diff.githubusercontent.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Cookie coming from patch-diff.githubusercontent.com attempted to set domain to github.com
Length: unspecified [text/plain]
Saving to: ‘2252.patch’

2252.patch                                     [ <=>                                                                                     ] 105.41K  --.-KB/s    in 0.05s

2022-12-09 10:57:02 (2.09 MB/s) - ‘2252.patch’ saved [107935]

[tkloczko@devel-g2v httpx]$ patch -p1 < 2252.patch
patching file httpx/_models.py
Reversed (or previously applied) patch detected!  Assume -R? [n]
```

**Comment by tomchristie** [comment]
GitHub's happy enough...

Screenshot 2022-12-09 at 11 05 03

We're blocked on an approving review.

**Review by kloczek — APPROVED** [review]
I cannot commit that PR on top of current master however if github all checks are passing ..

**Comment by tomchristie** [comment]
paging @encode/maintainers .

**Review by rafalp — APPROVED** [review]

**Comment by rafalp** [comment]
@tomchristie you've got it

**Comment by tomchristie** [comment]
Great thank you.
Let's aim to get a new minor release out first, and then merge this.

**Review by kloczek — APPROVED** [review]
LGTM.

**Review by kloczek — APPROVED** [review]

**Review by kloczek — APPROVED** [review]
LGTM

**Comment by kloczek** [comment]
Looks like PR is again out of date 😞

**Comment by kloczek** [comment]
Is it anytging else still on outstandung list related to this PR? 🤔

**Comment by Secrus** [comment]
is there any blocker stopping this from getting merged?

**Comment by tomchristie** [comment]
Looks ready to go from my side.
There's no breaking public API changes here, but since it removes a dependency I'd consider it as committing us to a proper version bump.

I've opened a discussion re. the next release here... https://github.com/encode/httpx/discussions/2534

**Comment by michealroberts** [comment]
@tomchristie How do we use this, in terms of using %20 over the + encoding ... ?

**Pull request merged**
